lib/pbench/server/api/resources/datasets_list.py

```diff
@@ -0,0 +1,148 @@
+from http import HTTPStatus
+from logging import Logger
+
+from flask.json import jsonify
+from flask.wrappers import Request, Response
+from flask_restful import abort
+
+from pbench.server import PbenchServerConfig
+from pbench.server.api.resources import (
+    ApiBase,
+    API_OPERATION,
+    JSON,
+    Parameter,
+    ParamType,
+    Schema,
+    SchemaError,
+)
+from pbench.server.database.database import Database
+from pbench.server.database.models.datasets import (
+    Dataset,
+    MetadataError,
+)
+
+
+class DatasetsList(ApiBase):
+    """
+    API class to list datasets based on PostgreSQL metadata
+    """
+
+    def __init__(self, config: PbenchServerConfig, logger: Logger):
+        super().__init__(
+            config,
+            logger,
+            Schema(
+                Parameter("name", ParamType.STRING),
+                Parameter("owner", ParamType.USER),
+                Parameter("access", ParamType.ACCESS),
+                Parameter("controller", ParamType.STRING),
+                Parameter("start", ParamType.DATE),
+                Parameter("end", ParamType.DATE),
+                Parameter(
+                    "metadata",
+                    ParamType.LIST,
+                    element_type=ParamType.KEYWORD,
+                    keywords=ApiBase.METADATA,
+                ),
+            ),
+            role=API_OPERATION.READ,
+        )
+
+    def _get(self, json_data: JSON, request: Request) -> Response:
+        """
+        Get a list of datasets matching a set of criteria.
+
+        NOTE: This does not rely on a JSON payload to identify the dataset and
+        desired metadata keys; instead we rely on URI query parameters.
+
+        TODO: The dispatcher is currently merging JSON request body and Flask
+        URI keywords ... it probably makes sense to also merge request query
+        parameters and then validate them all together; although it might make
+        more sense to validate each against separate schemas.
+
+        GET /api/v1/datasets/list?start=1970-01-01&end=2040-12-31&owner=fred&metadata=dashboard.seen&metadata=server.deletion
+        """
+
+        # We missed automatic schema validation due to the lack of a JSON body;
+        # construct an equivalent JSON body now so we can run it through the
+        # validator.
+        json = {
+            key: request.args.getlist(key)
+            if self.schema[key].type == ParamType.LIST
+            else request.args.get(key)
+            for key in request.args.keys()
+            if key in self.schema
```

Comment:
So this is a dictionary comprehension with an if/else statement.

How about using two helper functions?  That is a pain for humans to parse.  =)

def self._is_list(key):
    return self.schema[key].type == ParamType.LIST

def self._args_helper(key, args):
    return args.getlist(key) if self._is_list(key) else args.get(key)

json = {
    key: self._args_help(key, request.args) for key in request.args.keys() if key in self.schema
}